File: kernel/nodejs/src/bundle/source-staging.ts
/**
 * Fetching, extracting and writing the files a module's `sources:` block stages.
 *
 * One implementation for both callers: `telo release stage` stages every entry
 * of a module, and a kernel reading a source checkout stages the one entry a
 * resolution needs, on first use — the way it builds a controller from source.
 * Either way a file is written only once its bytes and execute bit match the
 * pin, so what lands on disk is what the manifest says it is.
 */

import {
  resolveSourceUrl,
  sourceUrlProblem,
  type ModuleSource,
  type SourceArchiveFormat,
  type SourceEntry,
} from "@telorun/analyzer";
import { NOOP_LOGGER, type Logger } from "@telorun/sdk";
import { createHash, randomBytes } from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { withDirectoryLock } from "../directory-lock.js";
import { assertPublicEgress } from "../transports/egress-guard.js";
import { checkStagedEntry, type StagedEntryState } from "./staged-entry.js";
import { readTarGz, type BundleEntry } from "./tar.js";

/** Reads the archive at a URL, in the format its source declares, into its entries. */
export type ArchiveReader = (url: string, format: SourceArchiveFormat) => Promise<BundleEntry[]>;

/** The archive could not be obtained: the network, the upstream, or the egress
 *  policy refused it. Distinct from an archive that was obtained and does not
 *  hold what the pin says, which is never a question of the environment. */
export class ArchiveFetchError extends Error {}

/** The archive was obtained and does not hold the file its entry declares: the
 *  member is absent or ambiguous, the archive does not read, or the bytes or
 *  execute bit differ from the pin. Any other staging failure — a lock that
 *  could not be taken, a write that failed — is neither this nor a fetch error. */
export class ArchiveContentError extends Error {}

/** A staging failure as the clause after "the file is staged by source 'x'",
 *  saying which kind it was, so a lock or a write is never read as a bad pin. */
export function describeStagingFailure(err: unknown): string {
  const detail = err instanceof Error ? err.message : String(err);
  if (err instanceof ArchiveFetchError) return `and its archive could not be fetched: ${detail}`;
  if (err instanceof ArchiveContentError) return `but its archive does not hold the pinned file: ${detail}`;
  return `and staging it failed: ${detail}`;
}

const FETCH_TIMEOUT_MS = 5 * 60_000;
const MAX_ARCHIVE_BYTES = 512 * 1024 * 1024;
const MAX_EXTRACTED_BYTES = 2 * 1024 * 1024 * 1024;
const MAX_REDIRECTS = 10;

function fetchFailure(url: string, err: unknown, timeoutMs: number): ArchiveFetchError {
  if ((err as { name?: unknown }).name === "TimeoutError") {
    return new ArchiveFetchError(
      `could not fetch the archive ${url}: no complete response within ${timeoutMs / 1000}s`,
    );
  }
  const cause = (err as { cause?: unknown }).cause;
  const detail = cause instanceof Error ? cause.message : err instanceof Error ? err.message : String(err);
  return new ArchiveFetchError(`could not fetch the archive ${url}: ${detail}`);
}

/**
 * One fetch per URL for the life of the reader; the caller decides that life. A
 * request is bounded in time, in response size and in decompressed size, may not
 * end at a URL a source could not name, and follows the kernel's egress policy
 * (`TELO_EGRESS`) at every hop.
 */
export function createArchiveReader(
  options: {
    fetchImpl?: typeof fetch;
    timeoutMs?: number;
    maxBytes?: number;
    maxExtractedBytes?: number;
  } = {},
): ArchiveReader {
  const fetchImpl = options.fetchImpl ?? fetch;
  const timeoutMs = options.timeoutMs ?? FETCH_TIMEOUT_MS;
  const maxBytes = options.maxBytes ?? MAX_ARCHIVE_BYTES;
  const maxExtractedBytes = options.maxExtractedBytes ?? MAX_EXTRACTED_BYTES;
  const cache = new Map<string, Promise<BundleEntry[]>>();
  return (url, format) => {
    const key = `${format}\0${url}`;
    let entries = cache.get(key);
    if (!entries) {
      entries = (async () => {
        const bytes = await fetchArchive(url, fetchImpl, timeoutMs, maxBytes);
        try {
          return await readTarGz(bytes, { maxBytes: maxExtractedBytes });
        } catch (err) {
          throw new ArchiveContentError(
            `the archive ${url} does not read as ${format}: ${err instanceof Error ? err.message : String(err)}`,
          );
        }
      })();
      cache.set(key, entries);
    }
    return entries;
  };
}

async function fetchArchive(
  url: string,
  fetchImpl: typeof fetch,
  timeoutMs: number,
  maxBytes: number,
): Promise<Buffer> {
  const tooLarge = () => new ArchiveFetchError(`the archive ${url} is larger than the ${maxBytes}-byte limit`);
  const signal = AbortSignal.timeout(timeoutMs);
  // Redirects are followed here rather than by fetch, so each location is vetted
  // before any request goes to it: a redirect may not take the request somewhere
  // the manifest could not have named — a plain-http location above all.
  let response: Response;
  let current = url;
  for (let hops = 0; ; hops++) {
    try {
      await assertPublicEgress(current);
    } catch (err) {
      throw new ArchiveFetchError(
        `the archive request ${current} was refused: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
    try {
      response = await fetchImpl(current, { signal, redirect: "manual" });
    } catch (err) {
      throw fetchFailure(url, err, timeoutMs);
    }
    const location = response.headers.get("location");
    if (response.status < 300 || response.status >= 400 || location === null) break;
    await response.body?.cancel();
    if (hops === MAX_REDIRECTS) {
      throw new ArchiveFetchError(`the archive request ${url} was redirected more than ${MAX_REDIRECTS} times`);
    }
    const next = new URL(location, current).href;
    const refusal = sourceUrlProblem(next);
    if (refusal) {
      throw new ArchiveFetchError(`the archive request ${url} was redirected to ${next}: ${refusal.detail}`);
    }
    current = next;
  }
  if (!response.ok) {
    await response.body?.cancel();
    throw new ArchiveFetchError(`the archive request ${url} answered ${response.status} ${response.statusText}`);
  }
  if (Number(response.headers.get("content-length") ?? 0) > maxBytes) {
    await response.body?.cancel();
    throw tooLarge();
  }
  const reader = response.body?.getReader();
  if (!reader) throw new ArchiveFetchError(`the archive request ${url} answered with no body`);
  const chunks: Buffer[] = [];
  let total = 0;
  for (;;) {
    let chunk: Awaited<ReturnType<typeof reader.read>>;
    try {
      chunk = await reader.read();
    } catch (err) {
      throw fetchFailure(url, err, timeoutMs);
    }
    if (chunk.done) break;
    total += chunk.value.byteLength;
    if (total > maxBytes) {
      await reader.cancel();
      throw tooLarge();
    }
    chunks.push(Buffer.from(chunk.value));
  }
  return Buffer.concat(chunks);
}

const sha256Hex = (bytes: Uint8Array): string => createHash("sha256").update(bytes).digest("hex");

const memberName = (name: string): string => name.replace(/^(\.\/)+/, "");

/** One file out of the archive a source names. */
export async function extractMember(
  archives: ArchiveReader,
  source: ModuleSource,
  url: string,
  member: string,
): Promise<{ content: Buffer; executable: boolean }> {
  const entries = await archives(url, source.archive);
  const matches = entries.filter((entry) => memberName(entry.name) === memberName(member));
  if (matches.length > 1) {
    throw new ArchiveContentError(
      `member '${member}' occurs ${matches.length} times in the archive, so which file it names is ambiguous`,
    );
  }
  const found = matches[0];
  if (!found) throw new ArchiveContentError(`member '${member}' is not in the archive`);
  if ("link" in found) {
    throw new ArchiveContentError(`member '${member}' is a symbolic link in the archive, not a file`);
  }
  const content = typeof found.content === "string" ? Buffer.from(found.content) : Buffer.from(found.content);
  return { content, executable: found.executable === true };
}

/**
 * Refuse an entry whose parent path crosses a symbolic link or a non-directory
 * on disk: the path is confined as text, but a checked-out link would carry
 * every rm, mkdir, write and symlink below it outside the module.
 */
function assertConfined(dir: string, entryPath: string): void {
  const segments = entryPath.split("/").slice(0, -1);
  let current = dir;
  for (const segment of segments) {
    current = path.join(current, segment);
    const stat = fs.lstatSync(current, { throwIfNoEntry: false });
    if (!stat) return;
    if (stat.isSymbolicLink()) {
      throw new Error(
        `'${path.relative(dir, current)}' is a symbolic link on disk; staging below it would write outside the module. Replace it with a directory.`,
      );
    }
    if (!stat.isDirectory()) {
      throw new Error(`'${path.relative(dir, current)}' exists on disk and is not a directory.`);
    }
  }
}

/** Replace `abs` in one step, so a concurrent reader sees the old file or the new
 *  one and never a partial write; a link left at the path is replaced, not
 *  written through. */
function replaceEntry(abs: string, write: (temp: string) => void): void {
  fs.mkdirSync(path.dirname(abs), { recursive: true });
  const temp = `${abs}.${process.pid}.${randomBytes(4).toString("hex")}.tmp`;
  try {
    write(temp);
    fs.renameSync(temp, abs);
  } catch (err) {
    fs.rmSync(temp, { force: true });
    throw err;
  }
}

/**
 * Fetch one file entry from its source's archive and write it, once its bytes
 * and execute bit match the pin. Throws naming what did not match; an archive
 * that could not be obtained throws {@link ArchiveFetchError}.
 */
async function stageFileEntry(
  dir: string,
  source: ModuleSource,
  entry: Extract<SourceEntry, { kind: "file" }>,
  archives: ArchiveReader,
): Promise<void> {
  if (!entry.pin) {
    throw new Error("the entry is not pinned — run `telo release stage --pin` to write its sha256 and executable");
  }
  assertConfined(dir, entry.path);
  const url = resolveSourceUrl(source, entry.upstream);
  const { content, executable } = await extractMember(archives, source, url, entry.member);
  const digest = sha256Hex(content);
  if (digest !== entry.pin.sha256) {
    throw new ArchiveContentError(
      `member '${entry.member}' hashes to sha256 ${digest}, but the pin is ${entry.pin.sha256}`,
    );
  }
  if (executable !== entry.pin.executable) {
    throw new ArchiveContentError(
      `member '${entry.member}' is ${executable ? "" : "not "}executable, but the pin says executable: ${entry.pin.executable}`,
    );
  }
  replaceEntry(path.join(dir, entry.path), (temp) => {
    fs.writeFileSync(temp, content);
    fs.chmodSync(temp, executable ? 0o755 : 0o644);
  });
}

/** Create one link entry, unless it is already stored as declared. */
function stageLinkEntry(dir: string, entry: Extract<SourceEntry, { kind: "link" }>): void {
  assertConfined(dir, entry.path);
  const abs = path.join(dir, entry.path);
  const existing = fs.lstatSync(abs, { throwIfNoEntry: false });
  if (existing?.isSymbolicLink() && fs.readlinkSync(abs) === entry.target) return;
  replaceEntry(abs, (temp) => fs.symlinkSync(entry.target, temp));
}

/** What a staged entry is once {@link ensureStagedEntry} returns: staging leaves
 *  nothing short of a match, and an unpinned entry is never staged. */
export type EnsuredEntryState = Extract<StagedEntryState, { state: "match" | "unpinned" }>;

export interface StagingOptions {
  /** One reader per resolution, so entries from one archive fetch it once. */
  readonly archives?: ArchiveReader;
  /** Receives one line when a fetch starts — a first run can wait minutes. */
  readonly log?: Logger;
}

/**
 * Bring one staged entry to its pin, staging it when it is missing or no longer
 * matches — following a link through its source to the file it leads to, and
 * creating the links on the way.
 *
 * An unpinned entry is returned as it is, never staged: nothing could verify what
 * a fetch returned. Concurrent callers, in this process or another, serialize per
 * archive URL, and each re-checks before fetching, so an archive is fetched once
 * while entries from different archives stage side by side.
 */
export async function ensureStagedEntry(
  dir: string,
  source: ModuleSource,
  entry: SourceEntry,
  options: StagingOptions = {},
): Promise<EnsuredEntryState> {
  const before = await checkStagedEntry(dir, source, entry);
  if (before.state === "match" || before.state === "unpinned") return before;

  const links: Extract<SourceEntry, { kind: "link" }>[] = [];
  let current: SourceEntry | undefined = entry;
  while (current?.kind === "link") {
    const link: Extract<SourceEntry, { kind: "link" }> = current;
    if (links.some((seen) => seen.path === link.path)) {
      throw new Error(`the link '${entry.path}' of source '${source.name}' leads back to '${link.path}'`);
    }
    links.push(link);
    current = source.entries.find((candidate) => candidate.path === link.resolved);
    if (!current) {
      throw new Error(`the link '${link.path}' of source '${source.name}' leads to no entry of that source`);
    }
  }
  const file = current;
  if (!file.pin) return { state: "unpinned" };
  const url = resolveSourceUrl(source, file.upstream);
  const archives = options.archives ?? createArchiveReader();
  const log = options.log ?? NOOP_LOGGER;

  // Under the module's own `.telo/`, which every checkout ignores, so a lock left
  // by a killed process never shows up as a change to the module.
  const lock = path.join(dir, ".telo", "staging", createHash("sha256").update(url).digest("hex").slice(0, 16));
  await withDirectoryLock(
    lock,
    "staged file",
    async () => {
      if ((await checkStagedEntry(dir, source, file)).state !== "match") {
        log.info(`staging '${file.path}' of the module at ${dir} from ${url}`, {
          "telo.staging.module": dir,
          "telo.staging.entry": file.path,
          "telo.staging.url": url,
        });
        await stageFileEntry(dir, source, file, archives);
      }
      for (const link of links.reverse()) stageLinkEntry(dir, link);
    },
    log,
  );
  return { state: "match" };
}

File: kernel/rust/src/bundle/source_staging.rs
//! Fetching, extracting and writing a file a module's `sources:` block stages, on
//! first use. Mirrors `../../../nodejs/src/bundle/source-staging.ts`: a file is
//! written only once its bytes and execute bit match the pin, and concurrent
//! kernels serialize per archive URL under the module's `.telo/staging` and
//! re-check before fetching. Staging every entry of a module (`telo release
//! stage`) is the Node CLI's, which this runtime does not carry.

use std::fs;
use std::io::{ErrorKind, Read};
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use sha2::{Digest, Sha256};
use telo_analyzer::source_entries::{
    resolve_source_url, source_url_problem, ModuleSource, SourceEntry, SourcePin,
};
use ureq::Agent;

use crate::bundle::files_integrity::PayloadFile;
use crate::bundle::staged_entry::{check_staged_entry, StagedEntryState};
use crate::bundle::tar::read_tar_gz_bounded;
use crate::directory_lock::{with_directory_lock, LockError};
use crate::transports::egress_guard::assert_public_egress;
use crate::transports::oci::oci_client::resolve_location;

const CONNECT_TIMEOUT: Duration = Duration::from_secs(30);
const FETCH_TIMEOUT: Duration = Duration::from_secs(5 * 60);
const MAX_ARCHIVE_BYTES: u64 = 512 * 1024 * 1024;
const MAX_EXTRACTED_BYTES: u64 = 2 * 1024 * 1024 * 1024;
const MAX_REDIRECTS: usize = 10;

#[derive(Debug, thiserror::Error)]
pub enum StagingError {
    /// The archive could not be obtained: the network, the upstream, or the
    /// egress policy refused it.
    #[error("{0}")]
    Fetch(String),
    /// The archive was obtained and does not hold the file its entry declares:
    /// the member is absent or ambiguous, the archive does not read, or the bytes
    /// or execute bit differ from the pin.
    #[error("{0}")]
    Content(String),
    /// Anything else: a lock that could not be taken, a write that failed, a
    /// path that would stage outside the module.
    #[error("{0}")]
    Failed(String),
}

impl StagingError {
    /// The clause after "the file is staged by source 'x'", saying which kind of
    /// failure it was, so a lock or a write is never read as a bad pin.
    pub fn describe(&self) -> String {
        match self {
            StagingError::Fetch(detail) => format!("and its archive could not be fetched: {detail}"),
            StagingError::Content(detail) => format!("but its archive does not hold the pinned file: {detail}"),
            StagingError::Failed(detail) => format!("and staging it failed: {detail}"),
        }
    }
}

impl From<LockError> for StagingError {
    fn from(err: LockError) -> Self {
        StagingError::Failed(err.to_string())
    }
}

fn io_failed(err: std::io::Error) -> StagingError {
    StagingError::Failed(err.to_string())
}

/// What a staged entry is once [`ensure_staged_entry`] returns: staging leaves
/// nothing short of a match, and an unpinned entry is never staged.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnsuredEntryState {
    Match,
    Unpinned,
}

/// GET `url`, following redirects here so each hop is vetted — by the egress
/// policy and by the rule a source's own url follows — and bounded in time and
/// size.
fn fetch_archive(url: &str) -> Result<Vec<u8>, StagingError> {
    let agent: Agent = Agent::config_builder()
        .http_status_as_error(false)
        .max_redirects(0)
        .timeout_connect(Some(CONNECT_TIMEOUT))
        .timeout_global(Some(FETCH_TIMEOUT))
        .build()
        .into();
    let mut current = url.to_string();
    for hop in 0..=MAX_REDIRECTS {
        assert_public_egress(&current).map_err(|err| {
            StagingError::Fetch(format!("the archive request {current} was refused: {err}"))
        })?;
        let mut response = agent
            .get(&current)
            .call()
            .map_err(|err| StagingError::Fetch(format!("could not fetch the archive {url}: {err}")))?;
        let status = response.status();
        let location = status
            .is_redirection()
            .then(|| response.headers().get("location"))
            .flatten()
            .and_then(|value| value.to_str().ok())
            .map(str::to_string);
        if let Some(location) = location {
            if hop == MAX_REDIRECTS {
                break;
            }
            let next = resolve_location(&current, &location);
            if let Some((_, detail)) = source_url_problem(&next) {
                return Err(StagingError::Fetch(format!(
                    "the archive request {url} was redirected to {next}: {detail}"
                )));
            }
            current = next;
            continue;
        }
        if !status.is_success() {
            return Err(StagingError::Fetch(format!(
                "the archive request {url} answered {} {}",
                status.as_u16(),
                status.canonical_reason().unwrap_or("")
            )));
        }
        let mut bytes = Vec::new();
        response
            .body_mut()
            .as_reader()
            .take(MAX_ARCHIVE_BYTES + 1)
            .read_to_end(&mut bytes)
            .map_err(|err| StagingError::Fetch(format!("could not fetch the archive {url}: {err}")))?;
        if bytes.len() as u64 > MAX_ARCHIVE_BYTES {
            return Err(StagingError::Fetch(format!(
                "the archive {url} is larger than the {MAX_ARCHIVE_BYTES}-byte limit"
            )));
        }
        return Ok(bytes);
    }
    Err(StagingError::Fetch(format!(
        "the archive request {url} was redirected more than {MAX_REDIRECTS} times"
    )))
}

fn member_name(name: &str) -> &str {
    let mut name = name;
    while let Some(rest) = name.strip_prefix("./") {
        name = rest;
    }
    name
}

/// One file out of an archive's entries.
fn extract_member(entries: Vec<PayloadFile>, member: &str) -> Result<(Vec<u8>, bool), StagingError> {
    let mut matches: Vec<PayloadFile> = entries
        .into_iter()
        .filter(|entry| member_name(entry.name()) == member_name(member))
        .collect();
    if matches.len() > 1 {
        return Err(StagingError::Content(format!(
            "member '{member}' occurs {} times in the archive, so which file it names is ambiguous",
            matches.len()
        )));
    }
    match matches.pop() {
        None => Err(StagingError::Content(format!("member '{member}' is not in the archive"))),
        Some(PayloadFile::Link { .. }) => Err(StagingError::Content(format!(
            "member '{member}' is a symbolic link in the archive, not a file"
        ))),
        Some(PayloadFile::Regular { content, executable, .. }) => Ok((content, executable)),
    }
}

/// Refuse an entry whose parent path crosses a symbolic link or a non-directory
/// on disk: the path is confined as text, but a checked-out link would carry the
/// write below it outside the module.
fn assert_confined(dir: &Path, entry_path: &str) -> Result<(), StagingError> {
    let segments: Vec<&str> = entry_path.split('/').collect();
    let mut current = dir.to_path_buf();
    for segment in &segments[..segments.len().saturating_sub(1)] {
        current.push(segment);
        let metadata = match fs::symlink_metadata(&current) {
            Ok(metadata) => metadata,
            Err(err) if err.kind() == ErrorKind::NotFound => return Ok(()),
            Err(err) => return Err(io_failed(err)),
        };
        let relative = current.strip_prefix(dir).unwrap_or(&current).display().to_string();
        if metadata.file_type().is_symlink() {
            return Err(StagingError::Failed(format!(
                "'{relative}' is a symbolic link on disk; staging below it would write outside the module. Replace it with a directory."
            )));
        }
        if !metadata.is_dir() {
            return Err(StagingError::Failed(format!(
                "'{relative}' exists on disk and is not a directory."
            )));
        }
    }
    Ok(())
}

/// Replace `abs` in one step, so a concurrent reader sees the old entry or the
/// new one and never a partial write.
fn replace_entry(abs: &Path, write: impl FnOnce(&Path) -> std::io::Result<()>) -> Result<(), StagingError> {
    if let Some(parent) = abs.parent() {
        fs::create_dir_all(parent).map_err(io_failed)?;
    }
    let nanos = SystemTime::now().duration_since(UNIX_EPOCH).map_or(0, |elapsed| elapsed.as_nanos());
    let temp = PathBuf::from(format!("{}.{}.{nanos}.tmp", abs.display(), std::process::id()));
    let written = write(&temp).and_then(|()| fs::rename(&temp, abs));
    if let Err(err) = written {
        let _ = fs::remove_file(&temp);
        return Err(io_failed(err));
    }
    Ok(())
}

#[cfg(unix)]
fn write_file(path: &Path, content: &[u8], executable: bool) -> std::io::Result<()> {
    use std::os::unix::fs::PermissionsExt;
    fs::write(path, content)?;
    fs::set_permissions(path, fs::Permissions::from_mode(if executable { 0o755 } else { 0o644 }))
}

#[cfg(windows)]
fn write_file(path: &Path, content: &[u8], _executable: bool) -> std::io::Result<()> {
    fs::write(path, content)
}

#[cfg(unix)]
fn write_link(target: &str, path: &Path) -> std::io::Result<()> {
    std::os::unix::fs::symlink(target, path)
}

#[cfg(windows)]
fn write_link(target: &str, path: &Path) -> std::io::Result<()> {
    std::os::windows::fs::symlink_file(target, path)
}

fn sha256_hex(bytes: &[u8]) -> String {
    Sha256::digest(bytes).iter().map(|byte| format!("{byte:02x}")).collect()
}

/// Fetch one file entry from its source's archive and write it, once its bytes
/// and execute bit match the pin.
fn stage_file_entry(
    dir: &Path,
    path: &str,
    url: &str,
    source: &ModuleSource,
    member: &str,
    pin: &SourcePin,
) -> Result<(), StagingError> {
    assert_confined(dir, path)?;
    let archive = fetch_archive(url)?;
    let entries = read_tar_gz_bounded(&archive, MAX_EXTRACTED_BYTES).map_err(|err| {
        StagingError::Content(format!("the archive {url} does not read as {}: {err}", source.archive))
    })?;
    let (content, executable) = extract_member(entries, member)?;
    let digest = sha256_hex(&content);
    if digest != pin.sha256 {
        return Err(StagingError::Content(format!(
            "member '{member}' hashes to sha256 {digest}, but the pin is {}",
            pin.sha256
        )));
    }
    if executable != pin.executable {
        return Err(StagingError::Content(format!(
            "member '{member}' is {}executable, but the pin says executable: {}",
            if executable { "" } else { "not " },
            pin.executable
        )));
    }
    replace_entry(&dir.join(path), |temp| write_file(temp, &content, executable))
}

fn stage_link_entry(dir: &Path, path: &str, target: &str) -> Result<(), StagingError> {
    assert_confined(dir, path)?;
    let abs = dir.join(path);
    let stored = fs::symlink_metadata(&abs)
        .ok()
        .filter(|metadata| metadata.file_type().is_symlink())
        .and_then(|_| fs::read_link(&abs).ok());
    if stored.is_some_and(|stored| stored.to_string_lossy().replace('\\', "/") == target) {
        return Ok(());
    }
    replace_entry(&abs, |temp| write_link(target, temp))
}

/// Bring one staged entry to its pin, staging it when it is missing or no longer
/// matches — following a link through its source to the file it leads to, and
/// creating the links on the way. An unpinned entry is returned as it is, never
/// staged. Concurrent callers serialize per archive URL, so an archive is fetched
/// once while entries from different archives stage side by side.
pub fn ensure_staged_entry(
    dir: &Path,
    source: &ModuleSource,
    entry: &SourceEntry,
) -> Result<EnsuredEntryState, StagingError> {
    match check_staged_entry(dir, source, entry).map_err(io_failed)? {
        StagedEntryState::Match => return Ok(EnsuredEntryState::Match),
        StagedEntryState::Unpinned => return Ok(EnsuredEntryState::Unpinned),
        StagedEntryState::Missing(_) | StagedEntryState::Mismatch(_) => {}
    }

    let mut links: Vec<(&str, &str)> = Vec::new();
    let mut current = entry;
    let (file, path, upstream, member, pin) = loop {
        match current {
            SourceEntry::File { path, upstream, member, pin, .. } => break (current, path, upstream, member, pin),
            SourceEntry::Link { path, target, resolved, .. } => {
                if links.iter().any(|(seen, _)| seen == path) {
                    return Err(StagingError::Failed(format!(
                        "the link '{}' of source '{}' leads back to '{path}'",
                        entry.path(),
                        source.name
                    )));
                }
                links.push((path, target));
                current = source.entries.iter().find(|candidate| candidate.path() == resolved).ok_or_else(|| {
                    StagingError::Failed(format!(
                        "the link '{path}' of source '{}' leads to no entry of that source",
                        source.name
                    ))
                })?;
            }
        }
    };
    let Some(pin) = pin else {
        return Ok(EnsuredEntryState::Unpinned);
    };
    let url = resolve_source_url(source, upstream);

    // Under the module's own `.telo/`, which every checkout ignores.
    let lock = dir.join(".telo").join("staging").join(&sha256_hex(url.as_bytes())[..16]);
    with_directory_lock(&lock, "staged file", || {
        if check_staged_entry(dir, source, file).map_err(io_failed)? != StagedEntryState::Match {
            eprintln!("telo: staging '{path}' of the module at {} from {url}", dir.display());
            stage_file_entry(dir, path, &url, source, member, pin)?;
        }
        for (path, target) in links.iter().rev() {
            stage_link_entry(dir, path, target)?;
        }
        Ok::<(), StagingError>(())
    })?;
    Ok(EnsuredEntryState::Match)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::io::Write;
    use std::net::TcpListener;
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Arc;
    use telo_analyzer::source_entries::read_module_sources;

    const LIB: &[u8] = b"\x7fELF library bytes";

    fn archive() -> Vec<u8> {
        let mut builder = tar::Builder::new(Vec::new());
        let mut header = tar::Header::new_gnu();
        header.set_size(LIB.len() as u64);
        header.set_mode(0o644);
        header.set_cksum();
        builder.append_data(&mut header, "package/lib/libx.so.1", LIB).unwrap();
        let tar = builder.into_inner().unwrap();
        let mut gz = flate2::write::GzEncoder::new(Vec::new(), flate2::Compression::default());
        gz.write_all(&tar).unwrap();
        gz.finish().unwrap()
    }

    /// Serves the archive at `/lib-1.0.0.tgz` and 404 elsewhere, counting requests.
    fn serve() -> (String, Arc<AtomicUsize>) {
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let base = format!("http://{}", listener.local_addr().unwrap());
        let requests = Arc::new(AtomicUsize::new(0));
        let counted = Arc::clone(&requests);
        let body = archive();
        std::thread::spawn(move || {
            for stream in listener.incoming() {
                let Ok(mut stream) = stream else { continue };
                let mut head = Vec::new();
                let mut byte = [0u8; 1];
                while !head.ends_with(b"\r\n\r\n") && stream.read(&mut byte).is_ok_and(|n| n == 1) {
                    head.push(byte[0]);
                }
                counted.fetch_add(1, Ordering::SeqCst);
                let found = String::from_utf8_lossy(&head).starts_with("GET /lib-1.0.0.tgz ");
                let (status, payload): (&str, &[u8]) = if found { ("200 OK", &body) } else { ("404 Not Found", b"") };
                let _ = write!(stream, "HTTP/1.1 {status}\r\nContent-Length: {}\r\nConnection: close\r\n\r\n", payload.len());
                let _ = stream.write_all(payload);
            }
        });
        (base, requests)
    }

    fn source(url: &str, sha256: &str) -> ModuleSource {
        let read = read_module_sources(&json!({ "sources": { "lib": {
            "version": "1.0.0", "url": url, "archive": "tar.gz", "notices": ["LICENSE"],
            "entries": {
                "native/libx.so.1": { "upstream": "x", "member": "package/lib/libx.so.1", "sha256": sha256, "executable": false },
                "native/libx.so": { "target": "libx.so.1" },
                "native/unpinned.so": { "upstream": "x", "member": "package/lib/libx.so.1" },
            },
        } } }));
        assert_eq!(read.problems, []);
        read.sources.into_iter().next().unwrap()
    }

    fn entry<'a>(source: &'a ModuleSource, path: &str) -> &'a SourceEntry {
        source.entries.iter().find(|candidate| candidate.path() == path).unwrap()
    }

    #[test]
    fn stages_a_missing_file_and_the_link_leading_to_it() {
        let (base, requests) = serve();
        let dir = tempfile::tempdir().unwrap();
        let lib = source(&format!("{base}/lib-{{version}}.tgz"), &sha256_hex(LIB));
        let state = ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so")).unwrap();
        assert_eq!(state, EnsuredEntryState::Match);
        assert_eq!(fs::read(dir.path().join("native/libx.so")).unwrap(), LIB);
        assert_eq!(requests.load(Ordering::SeqCst), 1);

        // Already matching: no second fetch.
        ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so.1")).unwrap();
        assert_eq!(requests.load(Ordering::SeqCst), 1);
    }

    #[test]
    fn stages_a_stale_file_again() {
        let (base, _) = serve();
        let dir = tempfile::tempdir().unwrap();
        fs::create_dir_all(dir.path().join("native")).unwrap();
        fs::write(dir.path().join("native/libx.so.1"), "an older release").unwrap();
        let lib = source(&format!("{base}/lib-{{version}}.tgz"), &sha256_hex(LIB));
        assert_eq!(
            ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so.1")).unwrap(),
            EnsuredEntryState::Match
        );
        assert_eq!(fs::read(dir.path().join("native/libx.so.1")).unwrap(), LIB);
    }

    #[test]
    fn refuses_bytes_that_do_not_match_the_pin_and_writes_nothing() {
        let (base, _) = serve();
        let dir = tempfile::tempdir().unwrap();
        let lib = source(&format!("{base}/lib-{{version}}.tgz"), &"0".repeat(64));
        match ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so.1")) {
            Err(StagingError::Content(detail)) => assert!(detail.contains("hashes to sha256"), "{detail}"),
            other => panic!("expected a pin mismatch, got {other:?}"),
        }
        assert!(!dir.path().join("native/libx.so.1").exists());
    }

    #[cfg(unix)]
    #[test]
    fn reports_a_failure_that_is_neither_a_fetch_nor_the_archive_content_as_neither() {
        let (base, requests) = serve();
        let dir = tempfile::tempdir().unwrap();
        let outside = tempfile::tempdir().unwrap();
        std::os::unix::fs::symlink(outside.path(), dir.path().join("native")).unwrap();
        let lib = source(&format!("{base}/lib-{{version}}.tgz"), &sha256_hex(LIB));
        match ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so.1")) {
            Err(StagingError::Failed(detail)) => assert!(detail.contains("'native' is a symbolic link"), "{detail}"),
            other => panic!("expected a staging failure, got {other:?}"),
        }
        assert_eq!(requests.load(Ordering::SeqCst), 0);
    }

    #[test]
    fn reports_an_upstream_that_cannot_provide_the_archive_as_a_fetch_failure() {
        let (base, _) = serve();
        let dir = tempfile::tempdir().unwrap();
        let lib = source(&format!("{base}/gone-{{version}}.tgz"), &sha256_hex(LIB));
        match ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/libx.so.1")) {
            Err(StagingError::Fetch(detail)) => assert!(detail.contains("gone-1.0.0.tgz answered 404"), "{detail}"),
            other => panic!("expected a fetch failure, got {other:?}"),
        }
    }

    #[test]
    fn returns_an_unpinned_entry_without_fetching_it() {
        let (base, requests) = serve();
        let dir = tempfile::tempdir().unwrap();
        let lib = source(&format!("{base}/lib-{{version}}.tgz"), &sha256_hex(LIB));
        assert_eq!(
            ensure_staged_entry(dir.path(), &lib, entry(&lib, "native/unpinned.so")).unwrap(),
            EnsuredEntryState::Unpinned
        );
        assert_eq!(requests.load(Ordering::SeqCst), 0);
    }
}
